### PR TITLE
Export history data

### PR DESCRIPTION
**Pull request by randrii**

### CODE REVIEW

File: src/main/java/com/rybka/dao/HibernateDAO.java [hunk 1]
```diff
@@ -26,15 +27,13 @@ public void save(CurrencyHistory currencyHistory) {
         }
     }
 
-    public void showTableRow() {
+    public List<CurrencyHistory> select() {
```
Comment (jarc0der) [inline src/main/java/com/rybka/dao/HibernateDAO.java]: bad method naming.

retrieveCurrencyHistory()
findCurrencyHistory()

File: src/main/java/com/rybka/service/CSVExportService.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.List;
+
+@Log4j
+public class CSVExportService {
+
+    public void export(List<CurrencyHistory> histories, String file) {
+        try (var fileWriter = new FileWriter(file); var csvWriter = new CSVWriter(fileWriter)) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/CSVExportService.java]: СSVWriter better to use as field. 
This method isn't testable.

Comment (jarc0der) [reply src/main/java/com/rybka/service/CSVExportService.java]: FileWriter as well.

File: src/main/java/com/rybka/service/CSVExportService.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.extern.log4j.Log4j;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.List;
+
+@Log4j
+public class CSVExportService {
+
+    public void export(List<CurrencyHistory> histories, String file) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/CSVExportService.java]: Did you consider to use interface at least for Csv/Json export services?

File: src/main/java/com/rybka/view/ExchangeView.java [hunk 4]
```diff
@@ -41,9 +50,24 @@ public void showDialog() {
                     total);
 
             hibernateDAO.save(convertedResult);
-            hibernateDAO.showTableRow();
+            var history = hibernateDAO.select();
 
             showExchange(convertedResult);
+
+            switch (userExportChoice) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/view/ExchangeView.java]: Let's look in the task description 
`According to documentation user should be able to export history of his exchanges that recorded to the DB.
Implement services that can export data to the CSV, JSON, Console.`

Why do you add this code?

File: src/main/java/com/rybka/view/ExchangeView.java [hunk 5]
```diff
@@ -28,9 +32,14 @@ public void showDialog() {
             System.out.print("Enter value: ");
             var userValue = scanner.nextDouble();
             System.out.print("Insert target currency: ");
-
             var userTargetCurrency = scanner.next().toUpperCase();
 
+            int userExportChoice;
+            do {
+                System.out.print("Choose way to export data (1-Console, 2-CSV file, 3-Json file): ");
```
Comment (jarc0der) [inline src/main/java/com/rybka/view/ExchangeView.java]: Why did you add this code? Take a look at task description.

File: src/main/java/com/rybka/service/JSONExportService.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    public JSONExportService(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
+    public void export(List<CurrencyHistory> histories, String fileName) {
+        try (FileWriter fileWriter = new FileWriter(fileName)) {
+            fileWriter.write(objectMapper.writeValueAsString(histories));
+            fileWriter.flush();
+        } catch (IOException e) {
+            log.error("Unable to export data to JSON. Reason: " + e.getMessage());
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/JSONExportService.java:32]: log and that's all? 
Maybe would be better to re thrown an exception?

File: src/main/java/com/rybka/view/ExchangeView.java [hunk 7]
```diff
@@ -41,9 +50,24 @@ public void showDialog() {
...
+
+            switch (userExportChoice) {
+                case 1:
+                    new ConsoleExportService().export(history);
+                    break;
+                case 2:
+                    new CSVExportService().export(history, "test.csv");
+                    break;
+                case 3:
+                    new JSONExportService(new ObjectMapper()).export(history, "test.json");
```
Comment (jarc0der) [inline src/main/java/com/rybka/view/ExchangeView.java]: Haven't you create instance of ObjectMapper earlier for the Connector service? Why you didn't reuse it?

File: src/main/java/com/rybka/service/CSVExportService.java (hunk 2 above)
Comment (jarc0der) [inline src/main/java/com/rybka/service/CSVExportService.java]: I believe CSVMapper(from jackson) is able to map and write list of data to the file. Check that please.

File: src/main/java/com/rybka/dao/HibernateDAO.java (hunk 1 above)
Comment (randrii) [reply src/main/java/com/rybka/dao/HibernateDAO.java]: fixed

File: src/main/java/com/rybka/service/CSVExportService.java (hunk 2 above)
Comment (randrii) [reply src/main/java/com/rybka/service/CSVExportService.java]: fixed

File: src/main/java/com/rybka/service/CSVExportService.java (hunk 3 above)
Comment (randrii) [reply src/main/java/com/rybka/service/CSVExportService.java]: fixed

File: src/main/java/com/rybka/view/ExchangeView.java (hunk 4 above)
Comment (randrii) [reply src/main/java/com/rybka/view/ExchangeView.java]: fixed

File: src/main/java/com/rybka/view/ExchangeView.java (hunk 5 above)
Comment (randrii) [reply src/main/java/com/rybka/view/ExchangeView.java]: fixed

File: src/main/java/com/rybka/service/JSONExportService.java (hunk 6 above)
Comment (randrii) [reply src/main/java/com/rybka/service/JSONExportService.java:32]: fixed

File: src/main/java/com/rybka/view/ExchangeView.java (hunk 7 above)
Comment (randrii) [reply src/main/java/com/rybka/view/ExchangeView.java]: fixed

File: src/main/java/com/rybka/service/CSVExportService.java (hunk 2 above)
Comment (randrii) [reply src/main/java/com/rybka/service/CSVExportService.java]: fixed

File: src/main/java/com/rybka/config/PropertyInfo.java [hunk 8]
```diff
@@ -1,8 +1,15 @@
 package com.rybka.config;
 
+import java.io.File;
+
 public final class PropertyInfo {
-    public static final String PROPERTY_FILE_NAME = "application.properties";
+    public static final String RESOURCE_PATH = "src" + File.separator + "main" + File.separator + "resources" + File.separator;
```
Comment (jarc0der) [inline src/main/java/com/rybka/config/PropertyInfo.java]: Don't think it's a good idea to store in this path. 

Let user configure path from export folder in application.properties.

File: src/main/java/com/rybka/config/PropertyInfo.java [hunk 9]
```diff
@@ -1,8 +1,15 @@
...
+    public static final String PROPERTY_FILE_NAME = RESOURCE_PATH + "application.properties";
     public static final String PROPERTY_EXCHANGE_SOURCE = "exchange.source";
     public static final String PROPERTY_EXCHANGE_KEY = "exchange";
     public static final String PROPERTY_PRIME_EXCHANGE_KEY = "prime_exchange";
+
+    public static final String FILE_NAME_PATTERN = "yyyyMMddhhmmss";
```
Comment (jarc0der) [inline src/main/java/com/rybka/config/PropertyInfo.java]: Pattern `yyyyMMddhhmmss` hard to read 20200220111213, use any separator to divide date and time.

File: src/main/java/com/rybka/service/CSVExportService.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        this.csvMapper = csvMapper;
+    }
+
+    public void export(List<CurrencyHistory> histories) {
+        csvMapper.disable(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY);
+        CsvSchema csvSchema = csvMapper.schemaFor(CurrencyHistory.class).withHeader();
+        ObjectWriter writer = csvMapper.writer(csvSchema.withLineSeparator("\n"));
+
+        try (var fileWriter = new FileWriter(PropertyInfo.RESOURCE_PATH
+                + new RandomNameService().generateFileName()
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/CSVExportService.java]: create method, there isn't any need to create a hole service. 
at least you can create FileUntils.class with static method.

File: src/main/java/com/rybka/service/JSONExportService.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+public class JSONExportService implements ExportService {
+    private final ObjectMapper objectMapper;
+
+    public JSONExportService(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
+    public void export(List<CurrencyHistory> histories) {
+        try (var fileWriter = new FileWriter(PropertyInfo.RESOURCE_PATH
+                + new RandomNameService().generateFileName()
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/JSONExportService.java]: generateFileName() could has parameters like generateFileName(fileExtention, prefixString)

File: src/main/java/com/rybka/config/PropertyInfo.java (hunk 9 above)
Comment (randrii) [reply src/main/java/com/rybka/config/PropertyInfo.java]: fixed

File: src/main/java/com/rybka/service/CSVExportService.java (hunk 10 above)
Comment (randrii) [reply src/main/java/com/rybka/service/CSVExportService.java]: fixed

File: src/main/java/com/rybka/config/PropertyInfo.java (hunk 8 above)
Comment (randrii) [reply src/main/java/com/rybka/config/PropertyInfo.java]: fixed

File: src/main/java/com/rybka/service/JSONExportService.java (hunk 11 above)
Comment (randrii) [reply src/main/java/com/rybka/service/JSONExportService.java]: fixed

File: src/main/java/com/rybka/config/FileUtils.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+public class FileUtils {
+
+    public static final String FILE_NAME_PATTERN = "yyyy-MM-dd_hh-mm-ss";
+    public static final String CSV_SUFFIX = ".csv";
+    public static final String JSON_SUFFIX = ".json";
+
+    public static String generateFileName(String fileSuffix) {
+        Date date = Calendar.getInstance().getTime();
+        DateFormat dateFormat = new SimpleDateFormat(FILE_NAME_PATTERN);
```
Comment (jarc0der) [inline src/main/java/com/rybka/config/FileUtils.java]: This field can be private static. No need to create dateFormat each time on method call.

Comment (jarc0der) [reply src/main/java/com/rybka/config/FileUtils.java]: We should consider the same for Date object. 
Personally, I prefer LocalDate/LocalDateTime from Java 8.

File: src/main/java/com/rybka/config/PropertyInfo.java [hunk 13]
```diff
@@ -1,7 +1,10 @@
 package com.rybka.config;
 
+import java.io.File;
+
 public final class PropertyInfo {
-    public static final String PROPERTY_FILE_NAME = "application.properties";
+    public static final String RESOURCE_PATH = "src" + File.separator + "main" + File.separator + "resources" + File.separator;
+    public static final String PROPERTY_FILE_NAME = RESOURCE_PATH + "application.properties";
```
Comment (jarc0der) [inline src/main/java/com/rybka/config/PropertyInfo.java]: Try to get rid of RESOURCE_PATH.

Comment (randrii) [reply src/main/java/com/rybka/config/PropertyInfo.java]: fixed

File: src/main/java/com/rybka/config/FileUtils.java (hunk 12 above)
Comment (randrii) [reply src/main/java/com/rybka/config/FileUtils.java]: fixed

File: src/main/java/com/rybka/config/FileUtils.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+public class FileUtils {
+
+    public static final String FILE_NAME_PATTERN = "yyyy-MM-dd_hh-mm-ss";
+    public static final String CSV_SUFFIX = ".csv";
+    public static final String JSON_SUFFIX = ".json";
+    private static final DateFormat DATE_FORMAT = new SimpleDateFormat(FILE_NAME_PATTERN);
+
+    public static String generateFileName(String fileSuffix) {
+        LocalDateTime DATE_TIME = LocalDateTime.now();
```
Comment (jarc0der) [inline src/main/java/com/rybka/config/FileUtils.java:15]: Why DATE_TIME is uppercase?

**Review by jarc0der — APPROVED** [review]

File: src/main/java/com/rybka/config/FileUtils.java (hunk 14 above)
Comment (randrii) [reply src/main/java/com/rybka/config/FileUtils.java:15]: fixed
